heat/core/linalg/bcg.py

```diff
@@ -100,59 +107,338 @@ def bi_diagonalize(A, overwrite_arr=True):
+                D1 = arr[i : i + 1 + bl, i]
+                v_left, tau_left = gen_house_vec(D1)
+                Uj = apply_house_left(D1, v_left, tau_left, U1, D1.shape[0], i)
+                arr[i : i + 1 + bl, i] = D1
+
+            if i < k - 1:
+                for j in range(2, k):
+                    # print("j is = ", j)
+
+                    if j == 2:
+                        if rank == 1:
+                            # print("here: ", arr)
+                            req3 = comm.irecv(source=0, tag=1)
+                            arr = req3.wait()
+                            # print("now here,", arr)
+                            Ej = arr[i : i + 1 + bl, i + 1 : i + b + 1]
+                            # print(Ej, end="    ")
+                            if Ej.size(0) > 0 and Ej.size(1) > 0:
+
+                                Uj = comm.recv(source=0, tag=2)
+                                Ej = torch.matmul(Uj.float(), Ej.float())
+                                # print("Ej is: ", Ej, end="  ")
+                                v_right, tau_right = gen_house_vec(Ej[0, :])
+                                vj = apply_house_right(Ej[0, :], v_right, tau_right, vt1, n, j)
+                                # arr[i,1:] = Ej[0,:]
+                                # print("Ej is: ", Ej)
+                                # print(i)
+                                arr[i : i + 1 + bl, i + 1 : i + b + 1] = Ej[:]
+                                # print("arr is: ", arr)
+                                # print(Ej[0, :])
+                                # print(arr[i,:])
+                                Aj = arr[i : i + bl + b + 1, i + 1 : i + b + 1]
+                                Aj = torch.matmul(Aj.float(), vj)
+                                # arr[i : i + bl + b + 1, i + 1 : i + b + 1] = Aj
+
+                            Dj = arr[i + 1 + bl : i + 1 + bl + b, i + 1 : i + 1 + b]
+                            if Dj.size(0) > 0 and Dj.size(1) > 0:
+                                v_left, tau_left = gen_house_vec(Dj[:, 0])
+                                Uj = apply_house_left(Dj, v_left, tau_left, U1, m, j)
+                                arr[i + 1 + bl : i + 1 + bl + b, i + 1 : i + 1 + b] = Dj
+                                # comm.send(Uj, dest=2)
+
+                            p_left, p_right = i + 1 + bl, i + 1 + bl + b
+                            # req = comm.isend(arr, dest=0, tag=3)
+                            # req.wait()
+                            # print(f"ok {i} and {j}", arr)
+
+                    else:
+                        # if rank == j - 1:
+                        # Uj = comm.recv(source=j - 1)
+                        if rank == 1:
+                            # print("This one:", arr)
+                            # req4 = comm.irecv(source=0,tag=1)
+                            # arr = req4.wait()
+                            Ej = arr[p_left:p_right, i + (j - 2) * b + 1 : i + 1 + (j - 1) * b]
```

Comment:
since you use this multiple times, it would probably help with readability if you saved `i + (j - 2) * b + 1` as a variable. the same goes for other slice start/stop indices